### ISSUE TITLE
train_on_batch outputs look wrong

### ISSUE THREAD

**Issue by htzh** [keras-team/keras#13559]
**System information**  
- Have I written custom code (as opposed to using example directory):  Yes
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04):  Google Colab default setting
- TensorFlow backend (yes / no):  yes
- TensorFlow version:  v1.15.0-0-g590d6eef7e 1.15.0
- Keras version:  
- Python version:  
- CUDA/cuDNN version:  
- GPU model and memory:  

**Describe the current behavior**  
If I run the following operations in the order provided:
```
(1) test_on_batch
(2) train_on_batch
(3) test_on_batch
```
The output from step (2) does not match those from either (1) or (3) and the differences are significant. I have checked that the ```test_on_batch``` output is as expected by using corresponding ```K.function``` to evaluate the losses. So I concluded that the ```train_on_batch``` output is quite misleading in showing what is going on during training. Of course I hope this does not affect the actual training updates but I don't have confidence and will need to figure out a way to check.

**Describe the expected behavior**  
Outputs from (1) and (2) should agree. But (2) agreeing with (3) is also acceptable.

**Code to reproduce the issue**  
I am working on a minimal example. 

* I have verified that in very simple models outputs from (1) and (2) do agree.

* My  loss functions involve gradients at the input. I have now removed all batch norm layers from my network (I only use batch size of 1). After the removal ```test_on_batch``` and ```train_on_batch``` agree on their first loss value (I have multiple loss values defined) and are different on the rest.  I have confirmed that ```train_on_batch``` kept weights unchanged from one loss valuation to the next for simple models.

* The story is getting more complicated. I have found that ```test_on_batch``` also produce wrong results involving gradients. But that probably belongs to a separate issue.

So it appears the mismatch happens only when batch norm is involved.
